src/Playwright.MSTest/PlaywrightTimeoutAttribute.cs

```diff
@@ -0,0 +1,115 @@
+        if (timeout == null || Debugger.IsAttached)
+        {
+            return base.Execute(testMethod);
+        }
+        if (CheckThatPlaywrightTimeoutIsSmallerThanNUnitTimeout(testMethod, timeout.Value) is var result && result != null)
+        {
+            return result;
+        }
+
+        return ExecuteWithTimeoutAsync(testMethod, timeout.Value).GetAwaiter().GetResult();
+    }
+
+    private async Task<TestResult[]> ExecuteWithTimeoutAsync(ITestMethod testMethod, int timeout)
+    {
+        // 1. Race Test execution against the specified timeout.
+        // In MSTest base.Execute will internally run the TestInitialize and TestCleanup. In case of a timeout, we don't
+        // run TestCleanup, only TestInitialize. To get good error messages, we close the browser context inside ContextCloseHookOnTimeout.
+        var testExecutionTask = Task.Run(() => base.Execute(testMethod));
+        var timeoutTask = Task.Delay(timeout);
+        await Task.WhenAny(testExecutionTask, timeoutTask);
+
+        // 2. Close the BrowserContext
+        if (ContextCloseHookOnTimeout != null)
+        {
+            await ContextCloseHookOnTimeout().ConfigureAwait(false);
+        }
+
+        // 3. If timeout was reached, we need to wait for the test execution to finish, since Page.Click could hang.
+        if (timeoutTask.IsCompleted)
+        {
+            // 3.1 Wait for the test execution to complete (let e.g. Page.ClickAsync throw)
+            // and give the test 5 seconds to finish.
+            await Task.WhenAny(testExecutionTask, Task.Delay(5000));
+        }
+
+        // 4. If the test execution is still running, we know it was a timeout.
+        if (!testExecutionTask.IsCompleted)
+        {
+            return new TestResult[]
+            {
+                new TestResult()
+                {
+                    Outcome = UnitTestOutcome.Timeout,
+                    TestFailureException = new Exception($"Test timed out after {timeout}ms.")
+                },
+            };
+        }
+
+        return await testExecutionTask;
+    }
+
+    private TestResult[]? CheckThatPlaywrightTimeoutIsSmallerThanNUnitTimeout(ITestMethod testMethod, int playwrightTestTimeout)
+    {
+        var timeout = MSTestProvider.TestTimeout;
+        var timeoutAttribute = testMethod.MethodInfo.GetCustomAttributes(typeof(TimeoutAttribute), false).FirstOrDefault();
+        if (timeoutAttribute != null)
+        {
+            timeout = ((TimeoutAttribute)timeoutAttribute).Timeout;
+        }
+
+        if (timeout != null && timeout != 0 && playwrightTestTimeout > timeout)
+        {
+            return new TestResult[]
+            {
+                new TestResult()
+                {
+                    Outcome = UnitTestOutcome.Failed,
+                    TestFailureException = new Exception($"Playwright test timeout ({playwrightTestTimeout}ms) is larger than MSTest timeout ({timeout}ms).")
```

Comment:
Maybe [InvalidOperationException](https://learn.microsoft.com/en-us/dotnet/api/system.invalidoperationexception) could be used here.

It would be nice to see the more verbose explanation you proposed above to know why Playwright timeout should be larger. I would also use longer rather than larger, as we are referring to time.